src/wasm/wat-parser.cpp

```diff
@@ -0,0 +1,1809 @@
+          lexer.take(1);
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  std::optional<uint64_t> takeU64() {
+    if (auto t = lexer.peek()) {
+      if (auto n = t->getU64()) {
+        lexer.take(1);
+        return n;
+      }
+    }
+    return {};
+  }
+
+  std::optional<int64_t> takeS64() {
+    if (auto t = lexer.peek()) {
+      if (auto n = t->getS64()) {
+        lexer.take(1);
+        return n;
+      }
+    }
+    return {};
+  }
+
+  std::optional<int64_t> takeI64() {
+    if (auto t = lexer.peek()) {
+      if (auto n = t->getI64()) {
+        lexer.take(1);
+        return n;
+      }
+    }
+    return {};
+  }
+
+  std::optional<uint32_t> takeU32() {
+    if (auto t = lexer.peek()) {
+      if (auto n = t->getU32()) {
+        lexer.take(1);
+        return n;
+      }
+    }
+    return {};
+  }
+
+  std::optional<int32_t> takeS32() {
+    if (auto t = lexer.peek()) {
+      if (auto n = t->getS32()) {
+        lexer.take(1);
+        return n;
+      }
+    }
+    return {};
+  }
+
+  std::optional<int32_t> takeI32() {
+    if (auto t = lexer.peek()) {
+      if (auto n = t->getI32()) {
+        lexer.take(1);
+        return n;
+      }
+    }
+    return {};
+  }
+
+  std::optional<double> takeF64() {
+    if (auto t = lexer.peek()) {
+      if (auto d = t->getF64()) {
+        lexer.take(1);
+        return d;
+      }
+    }
+    return {};
+  }
+
+  std::optional<float> takeF32() {
+    if (auto t = lexer.peek()) {
+      if (auto f = t->getF32()) {
+        lexer.take(1);
+        return f;
+      }
+    }
+    return {};
+  }
+
+  std::optional<std::string_view> takeString() {
+    if (auto t = lexer.peek()) {
+      if (auto s = t->getString()) {
+        lexer.take(1);
+        return s;
+      }
+    }
+    return {};
+  }
+
+  std::optional<Name> takeName() {
+    // TODO: Move this to lexer and validate UTF.
+    if (auto str = takeString()) {
+      // Copy to a std::string to make sure we have a null terminator, otherwise
+      // the `Name` constructor won't work correctly.
+      // TODO: Update `Name` to use string_view instead of char* and/or to take
+      // rvalue strings to avoid this extra copy.
+      return Name(std::string(*str));
+    }
+    return {};
+  }
+
+  struct Position {
+    const char* pos;
+  };
+
+  Position getPos() {
+    if (auto t = lexer.peek()) {
+      return {&t->span.front()};
+    }
+    return {&lexer.lexer.next().back()};
+  }
+
+  std::string_view getSpanSince(Position prev) {
+    auto curr = getPos();
+    return std::string_view(prev.pos, curr.pos - prev.pos);
+  }
+};
+
+// ===================
+// POD Utility Structs
+// ===================
+
+// The input span containing the definition for a module element and the
+// element's name, if it exists.
+struct DefSpan {
+  Name name;
+  std::string_view span;
+};
+
+struct Limits {
+  uint32_t min;
+  std::optional<uint32_t> max;
+};
+
+struct DefinedTableType {
```

Comment:
Slightly odd to have `DefSpan` but `DefinedTableType`, that is, to abbreviate Def sometimes but not others. Though I guess one is "definition" and one is "defined"? But, how about `TableInfo` here?

Reply:
Yeah, I would have preferred to call this just `TableType`, but then the compiler complains about the `using TableType = Ok;` below.

I'll solve this by changing the types in the `Ctx` types to be e.g. `using TableTypeT = Ok;`, `using TableTypeT = TableType`, etc.